base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -0,0 +1,259 @@
+            },
+        }
+    }
+
+    fn get_metadata(&self) -> Result<RwLockReadGuard<ChainMetadata>, ChainStorageError> {
+        self.metadata.read().map_err(|e| {
+            error!(
+                target: LOG_TARGET,
+                "An attempt to get sa read lock on the blockchain metadata failed. {}",
+                e.to_string()
+            );
+            ChainStorageError::AccessError("Read lock on blockchain metadata failed".into())
+        })
+    }
+
+    /// Returns the height of the current longest chain. This method will only fail if there's a fairly serious
+    /// synchronisation problem on the database. You can try calling [BlockchainDatabase::try_recover_metadata] in
+    /// that case to re-sync the metadata; or else just exit the program.
+    pub fn get_height(&self) -> Result<u64, ChainStorageError> {
+        let metadata = self.get_metadata()?;
+        Ok(metadata.height_of_longest_chain)
+    }
+
+    /// Returns the total accumulated work/difficulty of the longest chain.
+    ///
+    /// This method will only fail if there's a fairly serious synchronisation problem on the database. You can try
+    /// calling [BlockchainDatabase::try_recover_metadata] in that case to re-sync the metadata; or else
+    /// just exit the program.
+    pub fn get_total_work(&self) -> Result<Difficulty, ChainStorageError> {
+        let metadata = self.get_metadata()?;
+        Ok(metadata.greatest_accumulated_work.into())
+    }
+
+    /// Returns the transaction kernel with the given hash.
+    pub fn fetch_kernel(&self, hash: HashOutput) -> Result<Option<TransactionKernel>, ChainStorageError> {
+        fetch!(self, hash, TransactionKernel)
+    }
+
+    /// Returns the block header at the given block height.
+    pub fn fetch_header(&self, block_num: u64) -> Result<Option<BlockHeader>, ChainStorageError> {
+        fetch!(self, block_num, BlockHeader)
+    }
+
+    /// Returns the UTXO with the given hash.
+    pub fn fetch_utxo(&self, hash: HashOutput) -> Result<Option<TransactionOutput>, ChainStorageError> {
+        fetch!(self, hash, UnspentOutput)
+    }
+
+    /// Returns the STXO with the given hash.
+    pub fn fetch_stxo(&self, hash: HashOutput) -> Result<Option<TransactionOutput>, ChainStorageError> {
+        fetch!(self, hash, SpentOutput)
+    }
+
+    /// Returns the orphan block with the given hash.
+    pub fn fetch_orphan(&self, hash: HashOutput) -> Result<Option<Block>, ChainStorageError> {
+        fetch!(self, hash, OrphanBlock)
+    }
+
+    /// Returns true if the given UTXO, represented by its hash exists in the UTXO set.
+    pub fn is_utxo(&self, hash: HashOutput) -> Result<bool, ChainStorageError> {
+        let key = DbKey::UnspentOutput(hash);
+        self.db.contains(&key)
+    }
+
+    /// Calculate the Merklish root of the current UTXO set.
+    pub fn get_utxo_root(&self) -> Result<HashOutput, ChainStorageError> {
```

Comment:
```suggestion
    pub fn utxo_root(&self) -> Result<HashOutput, ChainStorageError> {
```
https://rust-lang-nursery.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

Reply:
Hmm, it's not a getter per se, Maybe `fetch_` or `calculate_`?

Reply:
So the main thing I was suggestion was to remove `get_` from all the getters in the Impl. Usually it's not such a big thing, but in this case there are also methods that start with `fetch_` so it is a little confusing when to use `fetch_` and when to use `get_`. I can see that fetch is most likely a query, where the `get`s have the data already available, but it might be confusing for consumers of this lib and future developers. I don't have a very strong opinion on this though, so we can leave it as is.

> You know, one thing we could do is drop this method and add MmrRoot values to the DbKey and DbValue enums

I actually like having the individual methods because they come up nicely in intellisense/code completion

Reply:
Ok, renamed to `fetch*`